### PR TITLE
Use aiter triton fused_add_rmsnorm_pad for gpt-oss

### PR DESCRIPTION
**Pull request by Rohan138**
## Purpose

Adds fused padding op before router GEMM on ROCm, eliminating this unfused pad after the GEMM before the fused_moe: https://github.com/ROCm/vllm/blob/main/vllm/model_executor/layers/fused_moe/layer.py#1603

Before:
image
After:
image

~Follow-up/alternate possibility is to replace this with a single `F.pad` before the router, then add a fusion pass to fuse AITER CK rmsnorm and pad to PassManager similar to https://github.com/vllm-project/vllm/pull/25693.~ Done

See also #30357 (gpt-oss quark w4a8 enablement) and #30647 (eliminate padding op on NV w4a8 gpt-oss)

## Test Result

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [ ] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [ ] The test plan, such as providing test command.
- [ ] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
- [ ] (Optional) Release notes update. If your change is user facing, please update the release notes draft in the [Google Doc](https://docs.google.com/document/d/1YyVqrgX4gHTtrstbq8oWUImOyPCKSGnJ7xtTpmXzlRs/edit?tab=t.0).
</details>

### CODE REVIEW

File: vllm/model_executor/models/gpt_oss.py [hunk 1]
```diff
@@ -233,7 +244,18 @@ def forward(
...
+                self.post_attention_layernorm.variance_epsilon,
+                residual,
+                x_pad_to_multiple=256,
+            )
+        else:
+            hidden_states, residual = self.post_attention_layernorm(
+                hidden_states, residual
+            )
         output = self.mlp(hidden_states)
         return output, residual
```
[the remark this answers is not in the document]
Comment (Rohan138) [reply vllm/model_executor/models/gpt_oss.py:238]: fused_add_rmsnorm_pad only pads the output tensor, not the residual. See https://github.com/ROCm/aiter/blob/main/aiter/ops/triton/fused_add_rmsnorm_pad.py and https://github.com/ROCm/aiter/blob/main/aiter/ops/triton/_triton_kernels/fused_add_rmsnorm_pad.py

File: vllm/model_executor/models/gpt_oss.py [hunk 2]
```diff
@@ -49,6 +50,9 @@
     maybe_prefix,
 )
 
+if current_platform.is_rocm() and rocm_aiter_ops.is_enabled():
```
Comment (gshtras) [inline vllm/model_executor/models/gpt_oss.py]: is_rocm is already a part of rocm_aiter_ops.is_enabled through @if_aiter_supported

File: vllm/model_executor/models/gpt_oss.py [hunk 3]
```diff
@@ -233,7 +237,18 @@ def forward(
         hidden_states = self.attn(hidden_states, positions)
 
         # Fully Connected
-        hidden_states, residual = self.post_attention_layernorm(hidden_states, residual)
+        if current_platform.is_rocm() and rocm_aiter_ops.is_enabled():
```
Comment (gshtras) [inline vllm/model_executor/models/gpt_oss.py]: Same as above. Also it's better to cache the aiter check in a class var

**Review by ProExpertProg — CHANGES_REQUESTED** [review]
Let's do this via compile pass instead of platform-specific model definition changes

File: vllm/compilation/pass_manager.py [hunk 4]
```diff
@@ -18,8 +18,9 @@
 
 if rocm_aiter_ops.is_enabled():
     from vllm.compilation.rocm_aiter_fusion import (
-        RocmAiterRMSNormFusionPass,
+        RocmAiterRMSNormQuantFusionPass,
```
Comment (Rohan138) [inline vllm/compilation/pass_manager.py:21]: Renamed to `RocmAiterRMSNormQuantFusionPass` for clarity

File: vllm/compilation/rocm_aiter_fusion.py [hunk 5]
```diff
@@ -399,3 +399,72 @@ def uuid(self) -> str:
...
+
+class AddAiterRMSNormPadPattern:
+    """
+    This pattern replaces aiter_rmsnorm_with_add & a pad op with a aiter_triton_add_rmsnorm_pad op.
+    """
+
+    AITER_TRITON_ADD_RMSNORM_PAD_OP = (
+        rocm_aiter_ops.get_aiter_triton_add_rmsnorm_pad_op()
+    )
+    PAD_SIZE = 192
```
Comment (Rohan138) [inline vllm/compilation/rocm_aiter_fusion.py]: This is currently hardcoded for the gpt-oss pad i.e. pad(2880, 256) = 2880 + 192 = 3072.

I can generalize this in a follow up, possibly in #32307 which will switch MI300 padding from 256 (back) to 128.

File: vllm/envs.py [hunk 6]
```diff
@@ -969,6 +970,11 @@ def _get_or_set_default() -> str:
     "VLLM_ROCM_USE_AITER_RMSNORM": lambda: (
         os.getenv("VLLM_ROCM_USE_AITER_RMSNORM", "True").lower() in ("true", "1")
     ),
+    # Whether to use fused add+rmsnorm+pad kernel for gpt-oss
+    "VLLM_ROCM_USE_AITER_TRITON_FUSED_ADD_RMSNORM_PAD": lambda: (
```
Comment (Rohan138) [inline vllm/envs.py]: Currently using the env var for testing, but I can drop it and keep it as default/stick it inside a pass_config pass instead.

File: vllm/compilation/pass_manager.py [hunk 7]
```diff
@@ -123,13 +124,16 @@ def configure(self, config: VllmConfig) -> None:
...
                     self.passes += [
-                        RocmAiterRMSNormFusionPass(config),
+                        RocmAiterRMSNormQuantFusionPass(config),
                     ]
             if self.pass_config.fuse_act_quant:
                 self.passes += [ActivationQuantFusionPass(config)]
                 if rocm_aiter_ops.is_enabled():
                     self.passes += [RocmAiterSiluMulFp8GroupQuantFusionPass(config)]
 
+            if rocm_aiter_ops.is_enabled():
```
Comment (ProExpertProg) [inline vllm/compilation/pass_manager.py]: Let's add a flag here (can be enabled by default in optimization level 1+)

File: vllm/compilation/rocm_aiter_fusion.py (hunk 5 above)
Comment (ProExpertProg) [reply vllm/compilation/rocm_aiter_fusion.py]: Can you make a pad size param like epsilon is?

File: vllm/model_executor/models/gpt_oss.py [hunk 8]
```diff
@@ -182,12 +182,21 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
             x = sequence_parallel_chunk(x)
 
         if current_platform.is_rocm():
+            # Pad before router GEMM to fuse with preceding aiter_rmsnorm_with_add
```
Comment (ProExpertProg) [inline vllm/model_executor/models/gpt_oss.py]: Instead of changing the model definition, the transformation should just transform `rms -> gemm -> pad` into `rms+pad -> gemm`.

Also cc @tjtanaa could we unify this path in general so that we can call `self.router` on ROCm as well?

File: vllm/compilation/rocm_aiter_fusion.py (hunk 5 above)
Comment (Rohan138) [reply vllm/compilation/rocm_aiter_fusion.py]: I made both `hidden_size` and `x_pad_to_multiple` params, then specified them in the [pass](https://github.com/ROCm/vllm/blob/6adba53bd0703ee2ce5333c3d99a0c9c4d9808cb/vllm/compilation/rocm_aiter_fusion.py#L463).

Alternatively, I could hardcode `hidden_size` in a get_inputs function instead:
```
    def get_inputs(self) -> list[torch.Tensor]:
        M, N = 5, 2880
        input = self.rmsnorm_matcher.empty(M, N) if self.rmsnorm_matcher.enabled else self.rmsnorm_matcher.empty_f32(M, N)
        weight = self.rmsnorm_matcher.empty(N)
        residual = self.rmsnorm_matcher.empty(M, N)
        return [input, weight, residual]

    def register(self, pm_pass: PatternMatcherPass) -> None:
        def pattern(
            input: torch.Tensor, weight: torch.Tensor, residual: torch.Tensor
        ) -> tuple[torch.Tensor, torch.Tensor]:
            _, N = input.shape
            pad_size = self.x_pad_to_multiple - N % self.x_pad_to_multiple
            result_rms, residual_out = self.rmsnorm_matcher(input, weight, residual)
            result = torch.nn.functional.pad(
                result_rms, (0, pad_size), mode="constant", value=0.0
            )
            return result, residual_out
```

File: vllm/compilation/pass_manager.py (hunk 7 above)
Comment (Rohan138) [reply vllm/compilation/pass_manager.py]: Added `fuse_act_padding`, please check if the conditions in `vllm/config/vllm.py` look good to you

File: vllm/model_executor/models/gpt_oss.py (hunk 8 above)
Comment (Rohan138) [reply vllm/model_executor/models/gpt_oss.py]: @ProExpertProg can I put both of those changes in a follow-up PR?
- Unify `rocm_unquantized_gemm` with `self.router` and remove the `F.pad` introduced here
- Benchmark `wvSplitKrc` vs `gemm_a16w16`
- switch the fusion to an `rmsnorm -> router_gemm -> pad` fusion pass

This PR is intended to be used with the AITER Triton GEMM (`VLLM_ROCM_USE_AITER_TRITON_GEMM=1`), we need some internal discussion+benchmarking to extend it to the splitK and splitKrc (#29843) GEMM paths in [rocm_unquantized_gemm_impl](https://github.com/vllm-project/vllm/blob/main/vllm/model_executor/layers/utils.py#L129)

Currently those calls reshape the input `(N,K) -> `, which results in this extra copy+fill between the rmsnorm and the GEMM:
image

For comparison, this is current main (AITER rmsnorm + fill + GEMM):
image

cc @dllehr-amd

Comment (ProExpertProg) [reply vllm/model_executor/models/gpt_oss.py]: Yeah I think the router unification should definitely be a follow-up. But let's fix the pattern now so we don't have to modify the model definition at all

Comment (Rohan138) [reply vllm/model_executor/models/gpt_oss.py]: @ProExpertProg put `rocm_unquantized_gemm` into the pattern`, PTAL

File: vllm/compilation/pass_manager.py (hunk 7 above)
Comment (Rohan138) [reply vllm/compilation/pass_manager.py]: cc @gshtras for viz on the new flag+conditions for it in `vllm/config/vllm.py`

File: tests/compile/test_fuse_act_padding.py [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+    ):
+        super().__init__()
+        self.norm = RMSNorm(hidden_size, eps=1e-5)
+        self.router = torch.nn.Linear(hidden_size, num_local_experts)
+        self.x_pad_to_multiple = x_pad_to_multiple
+        self.pad_dim = x_pad_to_multiple - (hidden_size % x_pad_to_multiple)
+
+    def forward(self, x, residual):
+        x, residual = self.norm(x, residual)
+        router_logits = rocm_unquantized_gemm(
```
Comment (ProExpertProg) [inline tests/compile/test_fuse_act_padding.py]: Let's replicate a few layers here to simulate residual reuse, as that can sometimes cause matching bugs.

**Review by ProExpertProg — APPROVED** [review]
LGTM, nice work! Just one comment about improving the test

File: tests/compile/test_fuse_act_padding.py (hunk 9 above)
Comment (Rohan138) [reply tests/compile/test_fuse_act_padding.py]: Done, checking CI failures
